### PR TITLE
Update etcd defrag and backup

### LINKED ISSUE #384: etcd defrag + backup: Avoid too many leader changes
As k8s cluster user, I want the k8s control plane to always be responsive, stable and safe.

We have a nightly job to defragment etcd and back it up on all control plane nodes, randomized a bit, so the defragmentation does not happen all at the same time.
This has been in existence for many months already, but due to a missing `--now` in `systemctl enable`, it has not really been active before.
As @matofeder [points out](https://github.com/SovereignCloudStack/k8s-cluster-api-provider/pull/355#issuecomment-1459872544), the defragmentation may block access to etcd for a while (seconds on typically sized etcd DBs), causing etcd leader changes (on multi-node etcd clusters) or temporary kube-api failures (on single-node etcd clusters).

Things to consider:
1. Possibly stronger protection from concurrent defragmentation on multiple etcd nodes by scheduling all etcd nodes from the leader instead of relying on the configured randomness in the timer start time.
2. Scheduling the leader etcd defragmentation last, as it will likely cause a leader change and we want to minimize these. (Starting with the leader would cause several leader changes ...)
3. Skipping the leader's defragmentation (for up to a week or infinitely?) to cause less leader changes?
4. Skipping defragmentation on single-node etcd installations?
5. Leaving this disabled for R4 and do more real-world tests before R5. (This is not without risk either, we have seen heavily fragmented etcds causing trouble in real-life already.)

### PR DESCRIPTION
**Pull request by matofeder**
This PR updates etcd defragmentation and backup script as follows:
- Script exits without any defragmentation/backup/trim action if:
  - It is executed on non leader etcd member
  - It is executed on etcd cluster with some unhealthy member
  - It is executed on single member etcd cluster

- Conditions above could be skipped and the script execution could be forced by the optional arguments:
  - `--force-single`
  - `--force-unhealthy`
  - `--force-nonleader`
 
- Script then defragments the etcd cluster as follows:
  - Defrag the non leader etcd members first
  - Change the leadership to the randomly selected and defragmentation completed etcd member
  - Defrag the local (ex-leader) etcd member

- Script then backup & trim local (ex-leader) etcd member

Note:
This script executes etcdctl commands like `etcdctl move-leader` or `etcdctl endpoint status --cluster` which were introduced in etcdctl version 3.3.0. The previous etcdctl client was installed as an `apt` package. The latest etcdctl version available in Ubuntu 20.04 repositories is v3.2.26, hence this PR also installs the desired version of etcdctl client. Etcdctl client is then used for etcd DB maintenance tasks.

Closes #384

### CODE REVIEW

File: terraform/files/template/cluster-template.yaml
Comment (matofeder) [inline terraform/files/template/cluster-template.yaml:113]: Note: $${#MEMBERS[@]} - double `$` here and below as a escape character for `clusterctl generate` command

**Comment by garloff** [comment]
We'll be doing this improvement past R4 (but likely pull it into the maintained branches).

**Comment by chess-knight** [comment]
I fixed the script in f5b897529f784543a4a808c88733a26a4a58aeb9. There was `{#` - jinja template comment, which cloud-init doesn't like. I also fixed the installation of *etcdctl* tool.

I have done multiple tests with this defrag script. For 1 control-plane and also for 3 control-plane cluster everything works as expected.

**Comment by garloff** [comment]
We do dev work on the `main` branch normally and can then (selectively) backport to maintained branches.

File: doc/Maintenance_and_Troubleshooting.md
Comment (joshmue) [inline doc/Maintenance_and_Troubleshooting.md:129]: > which causes the performance to decrease
> and may even cause premature out-of-space conditions

I think these claims should be backed by referring to etcd documentation or some case study.

[etcd defragmentation docs](https://etcd.io/docs/v3.5/op-guide/maintenance/#defragmentation) AFAIU only say that while defragmentation is useful for giving back disk space to the host filesystem, the space is still available to etcd - without notions about performance or out-of-space conditions.

Let's back this up with some references or remove it.

(Of course, doing defragmentation before doing a file-based backup would have advantages like smaller backup sizes)

**Comment by garloff** [comment]
> We do dev work on the `main` branch normally and can then (selectively) backport to maintained branches.

I have created `maintained/v5.0.x` and `maintained/v5.x` branches now.
But of course, we merge to main first and then, possibly even after some delay for real-world testing, merge the "backports" into maintained branches as we judge appropriate. (It's good for a PR against main to already mention that this might be material for backports -- cf. Cc: stable@kernel.org.)

File: doc/Maintenance_and_Troubleshooting.md
Comment (joshmue) [reply doc/Maintenance_and_Troubleshooting.md:129]: > If the defragmentation is not performed in a long run the out-of-space condition may occur on the host filesystem which may lead to performance degradation on both host and then potentially on etcd too. I did not find any shiny case study that shows the correlation between the host's out-of-space and the host's performance degradation as this could depend on many factors like the type of disk etc.

Well, if the disk really runs out of space, this may cause indeed `unpredictable etcd cluster behavior`, not just performance problems.

> It is true that defragmentation is useful for giving back disk space to the host filesystem. If the defragmentation is not performed in a long run the out-of-space condition may occur on the host filesystem

That implies that etcd does not reuse space for itself, causing a steady increase of disk space usage. That would be a contradiction to the content of the etcd documentation.

Under https://etcd.io/docs/v3.5/op-guide/maintenance/#space-quota, the docs say that defragmentation is required when the cluster ran out of space earlier - to bring the cluster in an operable state again.
But even if doing periodic defragmentation, there is AFAIU nothing preventing the keyspace to explode *between the defragmentation cycles*.

According to my understanding, reasonable sizings/rate-limits/compactions/quotas are required to prevent out-of-space conditions as well as possible; with defragmentation playing mostly a minor role when recovering/backing-up clusters.

Comment (matofeder) [reply doc/Maintenance_and_Troubleshooting.md]: I got your point that defragmentation does not prevent out-of-space conditions, but mainly compactions should be performed before actual defragmentation. Also, it could be unclear that etcd reuses space for itself. Therefore, I removed the statement from docs 41d003bba43d0362a93e369f911f7421769b3a8c.

Comment (joshmue) [reply doc/Maintenance_and_Troubleshooting.md]: Thanks!

> I got your point that defragmentation (alone) does not prevent out-of-space conditions, but compaction + defragmentation do this.

(Even though I'd still argue that AFAIU, defragmentation does not really prevent out-of-space conditions *at all*, unless other things on the same disk use space that should have been planned to be available to etcd)

Comment (garloff) [reply doc/Maintenance_and_Troubleshooting.md]: @joshmue - I agree that we should also start to look at rate limits ... if we want to substantially increase the operational safety.

**Review by garloff — COMMENTED** [review]
I will push the simplifications (removal of etcdctl choice) and the comments on cluster-admins NOT supposed to log into the k8s nodes into the branch.

File: doc/Maintenance_and_Troubleshooting.md
Comment (matofeder) [reply doc/Maintenance_and_Troubleshooting.md:160]: addressed by fixup 0526d5d68776e9351d18b123af0739b1c5dd700f

File: terraform/environments/environment-default.tfvars
Comment (matofeder) [reply terraform/environments/environment-default.tfvars]: addressed by fixup 1b6920adbb801403e70d9ae5906894064a0f613d

File: doc/configuration.md
Comment (matofeder) [reply doc/configuration.md]: addressed by fixup 1b6920adbb801403e70d9ae5906894064a0f613d

**Review by garloff — APPROVED** [review]
Looks good to me now, thanks.
PS: Let's test this in main for a while before backporting it to maintained/v5.x and v5.0.x.

**Comment by guettli** [comment]
FYI: There is a new tool for etc defragmentation: https://github.com/ahrtr/etcd-defrag  It is is written by a core maintainer of etcd.

**Pull request merged**
